python/ml4ir/base/features/feature_fns/categorical.py

```diff
@@ -0,0 +1,260 @@
+import tensorflow as tf
+from tensorflow.keras import layers
+from tensorflow import feature_column
+from tensorflow import lookup
+
+import copy
+
+from ml4ir.base.io import file_io
+
+
+def categorical_embedding_with_hash_buckets(feature_tensor, feature_info):
+    """
+    Convert a string feature tensor into a categorical embedding.
+    Works by first converting the string into num_hash_buckets buckets
+    each of size hash_bucket_size, then converting each hash bucket into
+    a categorical embdding of dimension embedding_size. Finally, these embeddings
+    are combined either through mean, sum or concat operations to generate the final
+    embedding.
```

Comment:
awesome, very small suggestion: [...] the final embedding + based on the feature_info + .